src/main/java/akka/stream/contrib/DirectoryChanges.java

```diff
@@ -0,0 +1,231 @@
+        private final Queue<Pair<Path, Change>> buffer = new ArrayDeque<>();
+        private final WatchService service = directoryPath.getFileSystem().newWatchService();
+        private final WatchKey watchKey = directoryPath.register(
+          service,
+          new WatchEvent.Kind<?>[] { ENTRY_CREATE, ENTRY_MODIFY, ENTRY_DELETE, OVERFLOW },
+          // this is com.sun internal, but the service is useless on OSX without it
+          SensitivityWatchEventModifier.HIGH
+        );
+
+        {
+          setHandler(out, new AbstractOutHandler(){
+
+            @Override
+            public void onPull() throws Exception {
+              if (!buffer.isEmpty()) {
+                pushHead();
+              } else {
+                doPoll();
+                if (!buffer.isEmpty()) {
+                  pushHead();
+                } else {
+                  schedulePoll();
+                }
+              }
+            }
+          });
+        }
+
+        @Override
+        public void onTimer(Object timerKey) {
+          if (!isClosed(out)) {
+            doPoll();
+            if (!buffer.isEmpty()) {
+              pushHead();
+            } else {
+              schedulePoll();
+            }
+          }
+        }
+
+        @Override
+        public void postStop() {
+          try {
+            if (watchKey.isValid()) watchKey.cancel();
+            service.close();
+          } catch (Exception ex) {
+            // Remove when #21168 is in a release
+            throw new RuntimeException(ex);
+          }
+        }
+
+        private void pushHead() {
+          final Pair<Path, Change> head = buffer.poll();
+          if (head != null) {
+            push(out, head);
+          }
+        }
+
+        private void schedulePoll() {
+          scheduleOnce("poll", pollInterval);
+        }
+
+        private void doPoll() {
+          try {
+            for (WatchEvent<?> event: watchKey.pollEvents()) {
+              final WatchEvent.Kind<?> kind = event.kind();
+
+              if (OVERFLOW.equals(kind)) {
+                // overflow means that some file system change events may have been missed,
+                // that may be ok for some scenarios but to make sure it does not pass unnoticed we fail the stage
+                failStage(new RuntimeException("Overflow from watch service: '" + directoryPath + "'"));
+
+              } else {
+                // if it's not an overflow it must be a Path event
+                @SuppressWarnings("unchecked")
+                final Path path = (Path) event.context();
+                final Path absolutePath = directoryPath.resolve(path);
+                final Change change = kindToChange(kind);
+
+                buffer.add(new Pair<>(absolutePath, change));
+                if (buffer.size() > maxBufferSize) {
+                  failStage(new RuntimeException("Max event buffer size " +
+                    maxBufferSize + " reached for $path"));
+                }
```

Comment:
When there is demand and the buffer is empty, so the case when it fails here is when there are more changes on one poll result than fits in the buffer.